### PR TITLE
Implemented fulfilled monitor

### PR DESCRIPTION
**Pull request by llvieira**
[GitLab task](https://git.lsd.ufcg.edu.br/chicog/fogbow/issues/501).

Hey guys, please review this new pull request.

### CODE REVIEW

File: src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            if (externalServiceAddresses != null) {
+                computeOrderInstance.setExternalServiceAddresses(externalServiceAddresses);
+            }
+        } catch (Throwable e) {
+            LOGGER.error("Error trying to get map of addresses (IP and Port) " +
+                    "of the compute instance for order: " + order, e);
+        }
+    }
+
+    public boolean isActiveConnectionFromInstance(ComputeOrderInstance computeOrderInstance) {
```
Comment (arn3tt) [inline src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java]: I think `isInstanceReachable` would be a better name for this method. Also, I think this method could be summarized to `return sshConnectivity.checkSSHConnectivity(computeOrderInstance)` or something like that.

File: src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java [hunk 2]
```diff
@@ -93,10 +93,10 @@ public void setSshExtraPorts(String sshExtraPorts) {
 	
 	public void setExternalServiceAddresses(Map<String, String> serviceAddresses) {
 		if (serviceAddresses != null) {
-			this.sshPublicAddress = serviceAddresses.get(CommonConfigurationConstants.SSH_SERVICE_NAME);
+            this.sshPublicAddress = serviceAddresses.get(CommonConfigurationConstants.SSH_SERVICE_NAME);
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java]: Please, be careful with the Intelijj indentation.

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+     * then order state is set to failed.
+     *
+     * @param order {@link Order}
+     * @throws OrderStateTransitionException Could not remove order from list of fulfilled orders.
+     */
+    protected void processFulfilledOrder(Order order) throws OrderStateTransitionException {
+        synchronized (order) {
+            OrderState orderState = order.getOrderState();
+
+            if (orderState.equals(OrderState.FULFILLED)) {
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: You are not testing when the FulfilledMonitor try to process an order in a state that is not FULFILLED.

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        Assert.assertNull(this.fulfilledOrderList.getNext());
+
+        Order test = this.failedOrderList.getNext();
+        assertNotNull(test);
+        Assert.assertEquals(order.getOrderInstance(), test.getOrderInstance());
+        Assert.assertEquals(OrderState.FAILED, test.getOrderState());
+    }
+
+    /**
+     * Test if a fulfilled order of an failed compute instance is
```
Comment (paulofelipefeitosa) [inline src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: _an active **remote** compute instance._

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        } else if (instanceState.equals(InstanceState.ACTIVE) && orderType.equals(OrderType.COMPUTE)) {
+            LOGGER.info("Processing active compute instance for order [" + order.getId() + "]");
+
+            ComputeOrderInstance computeOrderInstance = (ComputeOrderInstance) orderInstance;
+            this.computeInstanceConnectivity.setTunnelingServiceAddresses(order, computeOrderInstance);
+
+            if (!this.computeInstanceConnectivity.isActiveConnectionFromInstance(computeOrderInstance)) {
+                OrderStateTransitioner.transition(order, OrderState.FAILED);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import org.fogbowcloud.manager.core.models.orders.Order;
+import org.fogbowcloud.manager.core.models.orders.instances.ComputeOrderInstance;
+import org.fogbowcloud.manager.core.threads.FulfilledMonitor;
+
+import java.util.Map;
+
+/**
+ * Class to check SSH connectivity with an compute instance.
+ */
+public class ComputeInstanceConnectivityUtil {
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    private static final Logger LOGGER = Logger.getLogger(FulfilledMonitor.class);
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java]: Fixed!

File: src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java (hunk 1 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/utils/ComputeInstanceConnectivityUtil.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java (hunk 2 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/models/orders/instances/ComputeOrderInstance.java]: Okay!

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        Assert.assertNull(this.fulfilledOrderList.getNext());
+
+        Order test = this.failedOrderList.getNext();
+        assertNotNull(test);
+        Assert.assertEquals(order.getOrderInstance(), test.getOrderInstance());
+        Assert.assertEquals(OrderState.FAILED, test.getOrderState());
+    }
+
+    /**
+     * Test if a fulfilled order of an active compute instance is changed
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: Done!

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        this.thread.start();
+
+        Thread.sleep(500);
+
+        Assert.assertNotNull(this.fulfilledOrderList.getNext());
+        Assert.assertNull(this.failedOrderList.getNext());
+    }
+
+    /**
+     * Test if a fulfilled order of an active compute instance is changed
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: Done!

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        this.thread.start();
+
+        Thread.sleep(500);
+
+        Assert.assertNotNull(this.fulfilledOrderList.getNext());
+        Assert.assertNull(this.failedOrderList.getNext());
+    }
+
+    @Test
+    public void testRunProcessRemoteComputeOrderInstanceReachable() throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: Done!

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+    public void tearDown() {
+        if (this.thread != null) {
+            this.thread.interrupt();
+        }
+
+        super.tearDown();
+    }
+
+    /**
+     * Test if a fulfilled order of an active compute instance has not
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: Done!

File: src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java (hunk 4 above)
Comment (llvieira) [reply src/test/java/org/fogbowcloud/manager/core/threads/TestFulfilledMonitor.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        boolean isActive = true;
+
+        while (isActive) {
+            try {
+                Order order = this.fulfilledOrdersList.getNext();
+
+                if (order != null) {
+                    try {
+                        this.processFulfilledOrder(order);
+                    } catch (Throwable e) {
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+                } else {
+                    this.fulfilledOrdersList.resetPointer();
+                    LOGGER.info("There is no fulfilled order to be processed, sleeping for "
+                            + this.sleepTime + " milliseconds");
+                    Thread.sleep(this.sleepTime);
+                }
+            } catch (InterruptedException e) {
+                isActive = false;
+                LOGGER.warn("Thread interrupted", e);
+            } catch (Throwable e) {
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: Fixed!

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java (hunk 3 above)
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: Done!

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        InstanceState instanceState = orderInstance.getState();
+
+        if (instanceState.equals(InstanceState.FAILED)) {
+            LOGGER.info("Instance state is failed for order [" + order.getId() + "]");
+            OrderStateTransitioner.transition(order, OrderState.FAILED);
+        } else if (instanceState.equals(InstanceState.ACTIVE) && orderType.equals(OrderType.COMPUTE)) {
+            LOGGER.info("Processing active compute instance for order [" + order.getId() + "]");
+
+            ComputeOrderInstance computeOrderInstance = (ComputeOrderInstance) orderInstance;
+            this.computeInstanceConnectivity.setTunnelingServiceAddresses(order, computeOrderInstance);
```
[the remark this answers is not in the document]
Comment (llvieira) [reply src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: Ok, I removed this line.

File: src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+        InstanceState instanceState = orderInstance.getState();
+
+        if (instanceState.equals(InstanceState.FAILED)) {
+            LOGGER.info("Instance state is failed for order [" + order.getId() + "]");
+            OrderStateTransitioner.transition(order, OrderState.FAILED);
+        } else if (instanceState.equals(InstanceState.ACTIVE) && orderType.equals(OrderType.COMPUTE)) {
+            LOGGER.info("Processing active compute instance for order [" + order.getId() + "]");
+
+            ComputeOrderInstance computeOrderInstance = (ComputeOrderInstance) orderInstance;
```
Comment (paulofelipefeitosa) [inline src/main/java/org/fogbowcloud/manager/core/threads/FulfilledMonitor.java]: The compute order instance has not the tunneling service informations, who has this information is the compute order instance from the Order.

**Pull request merged**